crates/engine/tree/src/tree/payload_validator.rs

```diff
@@ -421,13 +421,54 @@ where
             handle.cache_metrics(),
         );
 
-        let output = if self.config.state_provider_metrics() {
-            let state_provider = InstrumentedStateProvider::from_state_provider(&state_provider);
-            let output = ensure_ok!(self.execute_block(&state_provider, env, &input, &mut handle));
-            state_provider.record_total_latency();
-            output
-        } else {
-            ensure_ok!(self.execute_block(&state_provider, env, &input, &mut handle))
+        // Execute the block and handle any execution errors
+        let output = {
+            let execution_result = if self.config.state_provider_metrics() {
+                let state_provider =
+                    InstrumentedStateProvider::from_state_provider(&state_provider);
+                let result = self.execute_block(&state_provider, env, &input, &mut handle);
+                state_provider.record_total_latency();
+                result
+            } else {
+                self.execute_block(&state_provider, env, &input, &mut handle)
+            };
+
+            match execution_result {
+                Ok(result) => result,
+                Err(execution_err) => {
```

Comment:
this function is already gigantic,

could we extract this error handling to a separate fn instead so that we can do

```
let Ok() else {return on_error}
```

Reply:
makes sense done here https://github.com/paradigmxyz/reth/pull/18356/commits/855a19d8044db838b125638ccc13444600b68392